src/Proto.Cluster/MemberList.cs

```diff
@@ -4,109 +4,149 @@
+        private static readonly object _lock = new object();
+        private static readonly Dictionary<string, MemberStatus> _members = new Dictionary<string, MemberStatus>();
+        private static readonly Dictionary<string, IMemberStrategy> _memberStrategyByKind = new Dictionary<string, IMemberStrategy>();
+
+        private static Subscription<object> _clusterTopologyEvnSub;
 
         internal static void SubscribeToEventStream()
         {
-            clusterTopologyEvnSub = Actor.EventStream.Subscribe<ClusterTopologyEvent>(Pid.Tell);
+            _clusterTopologyEvnSub = Actor.EventStream.Subscribe<ClusterTopologyEvent>(UpdateClusterTopology);
         }
 
         internal static void UnsubEventStream()
         {
-            Actor.EventStream.Unsubscribe(clusterTopologyEvnSub.Id);
-        }
-
-        internal static void Spawn()
-        {
-            Pid = Actor.SpawnNamed(Actor.FromProducer(() => new MemberListActor()), "memberlist");
+            Actor.EventStream.Unsubscribe(_clusterTopologyEvnSub.Id);
         }
 
-        internal static void Stop()
+        internal static string[] GetMembers(string kind)
         {
-            Pid.Stop();
+            return _memberStrategyByKind.TryGetValue(kind, out var memberStrategy)
+                       ? memberStrategy.GetAllMembers().FindAll(m => m.Alive).Select(m => m.Address).ToArray() : new string[0];
         }
 
-        public static async Task<string[]> GetMembersAsync(string kind)
+        internal static string GetPartition(string name, string kind)
         {
-            //if there are no nodes holding the requested kind, just wait
-            var res = await Pid.RequestAsync<MembersResponse>(new MembersByKindRequest(kind, true));
-            return res.Members;
+            return _memberStrategyByKind.TryGetValue(kind, out var memberStrategy)
+                       ? memberStrategy.GetPartition(name) : "";
         }
 
-        public static async Task<string> GetPartitionAsync(string name, string kind)
+        internal static string GetActivator(string kind)
         {
-            var res = await Pid.RequestAsync<MemberResponse>(new PartitionMemberRequest(name, kind));
-            return res.Address;
+            return _memberStrategyByKind.TryGetValue(kind, out var memberStrategy)
+                       ? memberStrategy.GetActivator() : "";
         }
 
-        public static async Task<string> GetActivatorAsync(string kind)
-        {
-            var res = await Pid.RequestAsync<MemberResponse>(new ActivatorMemberRequest(kind));
-            return res.Address;
-        }
-    }
-
-    internal class MembersByKindRequest
-    {
-        public MembersByKindRequest(string kind, bool onlyAlive)
+        internal static void UpdateClusterTopology(ClusterTopologyEvent msg)
         {
-            Kind = kind ?? throw new ArgumentNullException(nameof(kind));
-            OnlyAlive = onlyAlive;
+            lock (_lock)

```

Comment:
Replace with Monitor.TryEnter to allow timeouts IMO.
even if we just loop over that, at least we could send some notification that things are taking a long time